en/ros2/px4_ros2_msg_translation_node.md

```diff
@@ -0,0 +1,442 @@
+
+1. **Include New Headers in `all_translations.h`**
+
+   <!-- TODO: update GitHub URL -->
+
+   Add all newly created headers to [`translations/all_translations.h`](https://github.com/PX4/PX4-Autopilot/blob/message_versioning_and_translation/msg/translation_node/translations/all_translations.h) so that the translation node can find them.
+
+   For example, append the following line to `all_translation.h`:
+
+   ```c++
+   #include "translation_vehicle_attitude_v4.h"
+   ```
+
+Note that in the example above and in most cases, step 4 only requires the developer to create a direct translation for the definition change.
+This is because the changes only involved a single message.
+In more complex cases of splitting, merging and/or moving definitions then a generic translation must be created.
+
+For example when moving a field from one message to another, a single generic translation should be added with the two older message versions as input, and the two newer versions as output.
+This ensures there is no information lost when translating forward or backward.
+
+::: warning
+If a nested message definition changes, all messages including that message also require a version update.
+For example this would be the case for message [PositionSetpointTriplet](../msg_docs/PositionSetpointTriplet.md) if it were versioned.
+This is primarily important for services which are more likely reference other message definitions.
+:::
+
+## Implementation Details
+
+The translation node dynamically monitors the topics and services.
+It then instantiates the counterside of the publications and subscribers as required.
+For example if there is an external publisher for version 1 of a topic and subscriber for version 2.
+
+Internally, it maintains a graph of all known topic and version tuples (which are the graph nodes).
+The graph is connected by the message translations.
+As arbitrary message translations can be registered, the graph can have cycles and multiple paths from one node to another.
+Therefore on a topic update, the graph is traversed using a shortest path algorithm.
+When moving from one node to the next, the message translation method is called with the current topic data.
+If a node contains an instantiated publisher (because it previously detected an external subscriber), the data is published.
+Thus, multiple subscribers of any version of the topic can be updated with the correct version of the data.
+
+For translations with multiple input topics, the translation continues once all input messages are available.
+
+## Limitations
+
+- The current implementation depends on a service API that is not yet available in ROS Humble, and therefore does not support services when built for ROS Humble.
+- Services only support a linear history, i.e. no message splitting or merging.
```

Comment:
I think this means 
```suggestion
- Service messages only support a linear history, i.e. no message splitting or merging.
```

It matters because _services_ is a very overloaded term. We should at least make it specific so anyone who doesn't understand what a service message is (i.e. me) knows they can ignore this limitation.